feat: add delay queue and DLX (#10)
babylonchain/staking-queue-client: README.md

# staking-queue-client

> We focus on using rabbitMQ for message processing in Babylon today.

Reviewer: ```suggestion
This document focuses on the usage of RabbitMQ for message processing in the Babylon system.
```

- **Dead Letter Exchange (DLX)**: A special exchange used to reroute messages 
from main queues to delayed queues when processing fails or when a message needs to be delayed.

## Configuration Details

### Dead Letter Exchange (DLX)

- **Purpose**: The DLX handles messages that fail during processing or messages 
that need to be retried after a specific delay.
- **Name**: `common_dlx`

### Main Queues

Main queues are created by initiating the `NewQueueClient`.
For example, in the Babylon mainnet phase 1, we have following queues:

```go
const (
	ActiveStakingQueueName    string = "active_staking_queue"
	UnbondingStakingQueueName string = "unbonding_staking_queue"
	WithdrawStakingQueueName  string = "withdraw_staking_queue"
	ExpiredStakingQueueName   string = "expired_staking_queue"
)
```

- **Configuration**:
  - Each main queue is configured with the DLX (`common_dlx`) as the dead-letter exchange.
  - Messages that fail to process are sent to the DLX with a specific routing key 
  that directs them to the corresponding delayed queue. 
  The routing key is in the form of `{{queueName}}` + `_routing_key`

### Delayed Queues

> For each queue that are created, we also auto-provision a corresponding delayed queue.

Reviewer: ```suggestion
For each queue that is created, we also auto-provision a corresponding delayed queue.
```

The name of the delayed queue is in the form of `{{queueName}}` + `_delay`

- **Purpose**: Delayed queues hold messages for a predetermined time before 
they are sent back to the main queue for reprocessing.
- **TTL (Time To Live)**: Each delayed queue is configured with a TTL. 
After the TTL expires, messages are automatically sent back to the main queue.
- **Routing**: Messages are routed back to the main queue using the default exchange, 
which routes messages based on the queue name.

## Re-queueing Messages

> Re-queueing of messages occurs when our application is temporarily unable to process 

Reviewer: ```suggestion
Re-queueing of messages occurs when the application is temporarily unable to process 
```
